### PR TITLE
Ephemeris service

### PR DESCRIPTION
**Pull request by kaikreuzer**
Ported over from https://github.com/eclipse/smarthome/pull/6227.

@maggu2810 Did I do it correctly to add the jollyday lib to the compile- and runtime-bom?

Also-by: Gaël L'hopital <gael@lhopital.org>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
> @maggu2810 Did I do it correctly to add the jollyday lib to the compile- and runtime-bom?

Correction: ...to the compile-bom, runtime-bom AND the bundle pom itself?

File: bundles/org.openhab.core.ephemeris/pom.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  <name>openHAB Core :: Bundles :: Ephemeris</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.openhab.core.bundles</groupId>
+      <artifactId>org.openhab.core.config.core</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/pom.xml]: Please do not add any external dependency here.
That should be injected by the compile BOM already.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    private static final String CONFIG_DAYSET_WEEKEND = "weekend";
+    private static final String CONFIG_COUNTRY = "country";
+    private static final String CONFIG_REGION = "region";
+    private static final String CONFIG_CITY = "city";
+
+    private final Map<String, Set<DayOfWeek>> daysets = new HashMap<>();
+    private final Map<Object, HolidayManager> holidayManagers = new HashMap<>();
+    private final List<String> countryParameters = new ArrayList<>();
+
+    @NonNullByDefault({})
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: We place the nullness annotations between the visibility and the type.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    private String country;
+    @NonNullByDefault({})
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: We place the nullness annotations between the visibility and the type.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    protected void setLocaleProvider(LocaleProvider localeProvider) {
+        this.localeProvider = localeProvider;
+    }
+
+    protected void unsetLocaleProvider(LocaleProvider localeProvider) {
+        this.localeProvider = null;
+    }
+
+    @Override
+    @Nullable
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: We place the nullness annotations between the visibility and the type.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            URL url = new URL("file:" + filename);
+            Set<Holiday> days = getHolidayManager(url).getHolidays(date.toLocalDate(), date.toLocalDate());
+            return days.isEmpty() ? Optional.empty() : Optional.of(days.iterator().next().getPropertiesKey());
+        } else {
+            throw new FileNotFoundException();
+        }
+    }
+
+    @Override
+    @Nullable
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: We place the nullness annotations between the visibility and the type.

File: features/karaf/openhab-core/src/main/feature/feature.xml [hunk 6]
```diff
@@ -225,6 +215,17 @@
     <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.linuxsysfs/${project.version}</bundle>
   </feature>
 
+  <feature name="openhab-core-ephemeris" version="${project.version}">
+    <feature>openhab-core-base</feature>
+    <bundle>mvn:org.threeten/threeten-extra/1.4</bundle>
+    <bundle>mvn:de.jollyday/jollyday/0.5.7</bundle>
```
Comment (maggu2810) [inline features/karaf/openhab-core/src/main/feature/feature.xml]: For all external bundle dependencies (I identified we missed it for JNA on migration) we create an target platform feature that is referenced by the core features.

File: bundles/org.openhab.core.ephemeris/pom.xml (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.core.ephemeris/pom.xml]: That's what I expected as well. But as you can see [here](https://github.com/openhab/openhab-core/pull/506/commits), the PR build failed without it and once I added it, it succeeded. Any idea why?

**Comment by kaikreuzer** [comment]
@maggu2810 Any idea why travis failed? Did I do anything wrong wrt capability/requirement?

**Comment by kaikreuzer** [comment]
Thanks - hopefully it is fixed now.

File: bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/internal/engine/action/EphemerisActionService.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * @author Gaël L'hopital - Initial contribution and API
+ */
+@Component
+public class EphemerisActionService implements ActionService {
+
+    public static EphemerisManager ephemerisManager;
+
+    @Override
+    public String getActionClassName() {
+        return Ephemeris.class.getCanonicalName();
```
Comment (maggu2810) [inline bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/internal/engine/action/EphemerisActionService.java]: Wouldn't it make sense to add a default implementation to the `ActionService` that returns `getActionClass().getCanoncialName()`?

File: bundles/org.openhab.core.ephemeris/pom.xml (hunk 1 above)
Comment (maggu2810) [reply bundles/org.openhab.core.ephemeris/pom.xml]: On my machine it is working (building the project after removing that lines).

Looking at the check of the first commit:
* continuous-integration/travis-ci/pr: success
* PR-openHAB-Core: failure

I assume you should check your Jenkins configuration.

File: features/karaf/openhab-core/src/main/feature/feature.xml [hunk 8]
```diff
@@ -225,6 +215,17 @@
     <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.linuxsysfs/${project.version}</bundle>
   </feature>
 
+  <feature name="openhab-core-ephemeris" version="${project.version}">
+    <requirement>openhab.tp;filter:="(feature=jollyday)"</requirement>
+    <feature dependency="true">openhab.tp-jollyday</feature>
+    <feature>openhab-core-base</feature>
+    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.ephemeris/${project.version}</bundle>
+    <config name="org.openhab.ephemeris">
```
Comment (maggu2810) [inline features/karaf/openhab-core/src/main/feature/feature.xml]: Has `org.openhab.ephemeris` been choosen by intention (the `core` is missing between openhab and ephemeris)?

File: features/karaf/openhab-core/src/main/feature/feature.xml [hunk 9]
```diff
@@ -225,6 +215,17 @@
...
+        dayset-weekend=Saturday,Sunday
+        dayset-school=Monday,Tuesday,Thursday,Friday
```
[the remark this answers is not in the document]
Comment (maggu2810) [reply features/karaf/openhab-core/src/main/feature/feature.xml]: We could add modify the ephemeris service to use a predefined locale dependent configuration (using the locale provider) if there is no one explicit supplied. The list of supplied configurations per language could be extended over the time.

**Comment by svendiedrichsen** [comment]
@kaikreuzer I wrote the Jollyday library. If I can help out with it please contact me.

**Comment by meingraham** [comment]
Is there an ETA on a beta version of the Ephemeris service? I'm happy to be an early tester.

File: bundles/org.openhab.core.ephemeris/NOTICE [hunk 10]
```diff
@@ -0,0 +1,19 @@
+This content is produced and maintained by the Eclipse SmartHome project.
+
+* Project home: https://eclipse.org/smarthome/
```
Comment (wborn) [inline bundles/org.openhab.core.ephemeris/NOTICE]: Please update this to an openHAB NOTICE file.

**Comment by maggu2810** [comment]
@kaikreuzer Do you plan to update this PR?

**Comment by kaikreuzer** [comment]
I had put in on hold as long as I didn't have a proper IDE. Will try to find time to revisit it.

@clinique As this code was written by you, feel free to help getting it merged by doing PRs against this branch :-)

File: features/karaf/openhab-core/src/main/feature/feature.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply features/karaf/openhab-core/src/main/feature/feature.xml]: I'd suggest that @clinique can add this as a follow-up PR. For now, I have added Wednesday for the list as this should be the default for most countries.

File: features/karaf/openhab-core/src/main/feature/feature.xml (hunk 8 above)
Comment (kaikreuzer) [reply features/karaf/openhab-core/src/main/feature/feature.xml]: Yes, this is the standard scheme for any configuration settings - as a result, you can define the parameters in a file called `ephemeris.cfg` and don't need to deal with any namespaces.

File: bundles/org.openhab.core.ephemeris/NOTICE (hunk 10 above)
Comment (kaikreuzer) [reply bundles/org.openhab.core.ephemeris/NOTICE]: I have updated it to the openHAB notice, but as we have Jollyday only as a dependency, I don't think this needs to be added here - all code in this bundle is EPL.

**Comment by kaikreuzer** [comment]
All, sorry for the delay. All comments should be addressed now.

File: bundles/org.openhab.core.ephemeris/pom.xml (hunk 1 above)
Comment (maggu2810) [reply bundles/org.openhab.core.ephemeris/pom.xml]: IMHO my comment is still valid and should be addressed.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/EphemerisManager.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    @Nullable
+    String getBankHolidayName(int offset);
+
+    /**
+     * Get given day name from given userfile
+     *
+     * @param filename absolute or relative path to the file on local file system
+     * @param offset Today +/- offset days (+1 = tomorrow, -1 = yesterday)
+     * @return name of the day or null if no corresponding entry
+     * @throws FileNotFoundException or MalformedURLException
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/EphemerisManager.java]: If multiple exceptions are thrown you have to use multiple `@throws`.

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+import de.jollyday.ManagerParameter;
+import de.jollyday.ManagerParameters;
+
+/**
+ * This service provides functionality around ephemeris services and is the central service to be used directly by
+ * others.
+ *
+ * @author Gaël L'hopital - Initial contribution and API
+ */
+@Component(immediate = true, name = "org.openhab.ephemeris", property = {
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: Why immediate?

File: bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            countryParameters.add(config.get(CONFIG_REGION).toString());
+        }
+
+        if (config.containsKey(CONFIG_CITY)) {
+            countryParameters.add(config.get(CONFIG_CITY).toString());
+        }
+    }
+
+    @Reference
+    protected void setLocaleProvider(LocaleProvider localeProvider) {
```
Comment (maggu2810) [inline bundles/org.openhab.core.ephemeris/src/main/java/org/eclipse/smarthome/core/ephemeris/internal/EphemerisManagerImpl.java]: Can be simplified and the nulnness workaround an be removed if constructor injection is used (using constructor injection is not a must but I assume we all like to make the code better).

File: bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/internal/engine/action/EphemerisActionService.java (hunk 7 above)
Comment (maggu2810) [reply bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/internal/engine/action/EphemerisActionService.java]: No comment?

File: features/karaf/openhab-tp/src/main/feature/feature.xml [hunk 14]
```diff
@@ -119,10 +125,15 @@
 
   <feature name="openhab.tp-jmdns" description="An implementation of multi-cast DNS in Java." version="${project.version}">
     <capability>openhab.tp;feature=jmdns;version=3.5.5</capability>
-
     <bundle>mvn:org.jmdns/jmdns/3.5.5</bundle>
   </feature>
 
+  <feature name="openhab.tp-jollyday" description="Jollyday - A holiday API" version="${project.version}">
```
Comment (maggu2810) [inline features/karaf/openhab-tp/src/main/feature/feature.xml]: Why is this added twice (see 112-117 above)?

File: bundles/org.openhab.core.ephemeris/pom.xml (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.core.ephemeris/pom.xml]: Definitely - sorry, it slipped through. Fixed now.

**Comment by kaikreuzer** [comment]
Thanks, @maggu2810, I have addressed all your comments.

File: bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/actions/Ephemeris.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    @ActionDoc(text = "get holiday for today from a given userfile")
+    public static String getBankHolidayName(String filename) {
+        return getBankHolidayName(0, filename);
+    }
+
+    @ActionDoc(text = "get holiday for today plus or minus an offset from a given userfile")
+    public static String getBankHolidayName(int offset, String filename) {
+        try {
+            return EphemerisActionService.ephemerisManager.getBankHolidayName(offset, filename);
+        } catch (MalformedURLException | FileNotFoundException e) {
```
Comment (maggu2810) [inline bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/actions/Ephemeris.java]: Travis is not happy because you catch `MalformedURLException` here but that exception is not thrown anymore.

```
[ERROR] /home/travis/build/openhab/openhab-core/bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/actions/Ephemeris.java:[81,3060] Unreachable catch block for MalformedURLException. This exception is never thrown from the try statement body
```

Comment (kaikreuzer) [reply bundles/org.openhab.core.model.script/src/org/eclipse/smarthome/model/script/actions/Ephemeris.java]: My IDE didn't show it, so I missed it. Fixed now.

**Comment by kaikreuzer** [comment]
Seems I still got something wrong wrt to the feature definitions...

File: features/karaf/openhab-core/src/main/feature/feature.xml [hunk 16]
```diff
@@ -114,6 +114,16 @@
     <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.linuxsysfs/${project.version}</bundle>
   </feature>
 
+  <feature name="openhab-core-ephemeris" version="${project.version}">
```
Comment (maggu2810) [inline features/karaf/openhab-core/src/main/feature/feature.xml]: You miss the ephemeris bundle itself in this very specific feature

Comment (kaikreuzer) [reply features/karaf/openhab-core/src/main/feature/feature.xml]: Good point. No idea how that got lost...

**Comment by maggu2810** [comment]
I did not realize the namesoace on review.
Why do we use `org.eclipse.smarthome` for code that is newly added to the repository?

**Comment by maggu2810** [comment]
@kaikreuzer Can you comment?

**Comment by kaikreuzer** [comment]
> Why do we use org.eclipse.smarthome for code that is newly added to the repository?

Because of https://github.com/openhab/openhab-core/pull/506#issue-249546600: "Ported over from eclipse/smarthome#6227". This was a contribution to ESH, which just wasn't merged on time.
I agree that we could have adapted the namespace, but I tried to keep it the original contribution that it was.
Ok for me to do a follow-up PR with an updated namespace, if you wish. Otherwise, we do it with an upcoming bulk rename for 3.x.

**Comment by maggu2810** [comment]
We renamed the whole automation part that has already been part of ESH.
I don't understand for which part it is important and for which contribution it is optional.

**Comment by kaikreuzer** [comment]
For automation it was exceptionally important as this code was a new API that many bindings immediately wanted to use, so a lot of code was implemented using it.
